File: fp_examples/compose.js
// Arg first composition

const composeFactory = (arg) =>
	(...fns) =>
		fns.reduce((acc, fn) => fn(acc), arg);


const concat = (append) => (str) => str + append;
const repeat = (times) => (str) => str.repeat(times);
const split = (by) => (str) => str.split(by);

const compose = composeFactory('Hello World');
const result = compose(
	concat('!'),
  	repeat(2),
  	split(' ')
);

console.log(result);


// Function first composition

const composer = (...fns) =>
	(arg) =>
		fns.reduce((acc, fn) => fn(acc), arg);

const processText = composer(
	concat('!'),
  	repeat(2),
  	split(' ')
);

console.log(processText('Hello Again'));
		

File: fp_examples/compose.py
from functools import reduce

def compose(*fns):
    return lambda arg: reduce(lambda v, f: f(v), fns, arg)

concat = lambda append: lambda str: str + append
repeat = lambda times: lambda str: str * times
remove = lambda repl: lambda str: str.replace(repl, '')

process_text = compose(
    concat("World!"),
    repeat(3),
    remove("!")
)

print(process_text('Hello Again'))